usr/man/cat1/curses.1



CURSES(3V)             C LIBRARY FUNCTIONS             CURSES(3V)



NAME
     curses - System V terminal screen handling and  optimization
     package

SYNOPSIS
     The curses manual page is organized as follows:

     In SYNOPSIS

          o+  compiling information

          o+  summary of parameters used by curses routines

     In SYSTEM V SYNOPSIS:

          o+  compiling information

     In DESCRIPTION and SYSTEM V DESCRIPTION:

          o+  An overview of how curses routines should be used

     In ROUTINES, descriptions of  curses  routines  are  grouped
     under the appropriate topics:

          o+  Overall Screen Manipulation

          o+  Window and Pad Manipulation

          o+  Output

          o+  Input

          o+  Output Options Setting

          o+  Input Options Setting

          o+  Environment Queries

          o+  Low-level Curses Access

          o+  Miscellaneous

          o+  Use of curscr

     In SYSTEM V ROUTINES, descriptions of  curses  routines  are
     grouped under the appropriate topics:

          o+  Overall Screen Manipulation

          o+  Window and Pad Manipulation

          o+  Output



Sun Release 4.1   Last change: 21 January 1990                  1






CURSES(3V)             C LIBRARY FUNCTIONS             CURSES(3V)



          o+  Input

          o+  Output Options Setting

          o+  Input Options Setting

          o+  Environment Queries

          o+  Soft Labels

          o+  Low-level Curses Access

          o+  Terminfo-Level Manipulations

          o+  Termcap Emulation

          o+  Miscellaneous

          o+  Use of curscr

     Then come sections on:

          o+  SYSTEM V ATTRIBUTES

          o+  SYSTEM V FUNCTION KEYS

          o+  LINE GRAPHICS

     cc [ _f_l_a_g_s ] _f_i_l_e_s -lcurses -ltermcap [ _l_i_b_r_a_r_i_e_s ]

     #include <curses.h>      (automatically  includes  <stdio.h>
                              and <unctl.h>.)

     The parameters in the following list are  not  global  vari-
     ables.   This  is  a  summary  of the parameters used by the
     curses library routines.  All routines return the int values
     ERR  or  OK  unless  otherwise  noted.  Routines that return
     pointers always return NULL on error.  ERR, OK  ,  and  NULL
     are   all  defined  in  <curses.h>.)  Routines  that  return
     integers are not listed in the parameter list below.

     bool bf
     char **area,*boolnames[], *boolcodes[], *boolfnames[], *bp
     char *cap, *capname, codename[2], erasechar, *filename, *fmt
     char *keyname, killchar, *label, *longname
     char *name, *numnames[], *numcodes[], *numfnames[]
     char *slk_label, *str, *strnames[], *strcodes[], *strfnames[]
     char *term, *tgetstr, *tigetstr, *tgoto, *tparm, *type
     chtype attrs, ch, horch, vertch
     FILE *infd, *outfd
     int begin_x, begin_y, begline, bot, c, col, count
     int dmaxcol, dmaxrow, dmincol, dminrow, *errret, fildes



Sun Release 4.1   Last change: 21 January 1990                  2






CURSES(3V)             C LIBRARY FUNCTIONS             CURSES(3V)



     int (*init()), labfmt, labnum, line
     int ms, ncols, new, newcol, newrow, nlines, numlines
     int oldcol, oldrow, overlay
     int p1, p2, p9, pmincol, pminrow, (*putc()), row
     int smaxcol, smaxrow, smincol, sminrow, start
     int tenths, top, visibility, x, y
     SCREEN *new, *newterm, *set_term
     TERMINAL *cur_term, *nterm, *oterm
     va_list varglist
     WINDOW *curscr, *dstwin, *initscr, *newpad, *newwin, *orig
     WINDOW *pad, *srcwin, *stdscr, *subpad, *subwin, *win

SYSTEM V SYNOPSIS
     /usr/5bin/cc [ _f_l_a_g ...] _f_i_l_e ...  -lcurses [ _l_i_b_r_a_r_y ...]

     #include <curses.h>      (automatically includes  <stdio.h>,
                              <termio.h>, and <unctrl.h>).

DESCRIPTION
     These routines give the user a method  of  updating  screens
     with  reasonable  optimization.   They  keep an image of the
     current screen, and the user sets up an image of a new  one.
     Then  the  refresh()  tells the routines to make the current
     screen look like the new one.  In order  to  initialize  the
     routines, the routine initscr() must be called before any of
     the other routines that deal with windows  and  screens  are
     used.  The routine endwin() should be called before exiting.

SYSTEM V DESCRIPTION
     The curses routines give  the  user  a  terminal-independent
     method of updating screens with reasonable optimization.

     In order to initialize the routines, the  routine  initscr()
     or newterm() must be called before any of the other routines
     that deal with windows and screens are used.   Three  excep-
     tions are noted where they apply.  The routine endwin() must
     be called before exiting.  To get character-at-a-time  input
     without echoing, (most interactive, screen oriented programs
     want this) after calling initscr() you should  call  `cbreak
     ();  noecho ();' Most programs would additionally call `nonl
     (); intrflush(stdscr, FALSE); keypad(stdscr, TRUE);'.

     Before a curses program  is  run,  a  terminal's  TAB  stops
     should  be  set  and its initialization strings, if defined,
     must be output.  This can be done by executing the tset com-
     mand  in your .profile or .login file.  For further details,
     see tset(1) and the Tabs and  Initialization  subsection  of
     terminfo(5V).

     The curses library contains routines  that  manipulate  data
     structures  called  _w_i_n_d_o_w_s  that  can be thought of as two-
     dimensional arrays of characters representing all or part of



Sun Release 4.1   Last change: 21 January 1990                  3






CURSES(3V)             C LIBRARY FUNCTIONS             CURSES(3V)



     a  terminal  screen.  A default window called stdscr is sup-
     plied, which is the size of the terminal screen.  Others may
     be  created with newwin().  Windows are referred to by vari-
     ables declared as WINDOW *; the type WINDOW  is  defined  in
     <curses.h>  to  be a C structure.  These data structures are
     manipulated with routines described below, among  which  the
     most basic are move() and addch().  More general versions of
     these routines are included with  names  beginning  with  w,
     allowing  you  to specify a window.  The routines not begin-
     ning with  w  usually  affect  stdscr.   Then  refresh()  is
     called,  telling  the  routines  to make the user's terminal
     screen look like stdscr.  The characters  in  a  window  are
     actually of type chtype, so that other information about the
     character may also be stored with each character.

     Special windows called _p_a_d_s may also be manipulated.   These
     are  windows  that  are  not  constrained to the size of the
     screen and whose contents need not be displayed  completely.
     See the description of newpad() under Window and Pad Manipu-
     lation for more information.

     In addition to  drawing  characters  on  the  screen,  video
     attributes may be included that cause the characters to show
     up in modes such as underlined or in reverse video on termi-
     nals  that  support such display enhancements.  Line drawing
     characters may be specified to be output.  On input,  curses
     is  also  able  to  translate  arrow  and function keys that
     transmit escape sequences into  single  values.   The  video
     attributes,  line  drawing  characters, and input values use
     names, defined in <curses.h>, such as A_REVERSE,  ACS_HLINE,
     and KEY_LEFT.

     curses also defines the WINDOW * variable, curscr, which  is
     used only for certain low-level operations like clearing and
     redrawing a garbaged screen.  curscr can be used in  only  a
     few  routines.   If  the  window  argument  to  clearok() is
     curscr, the next call to wrefresh()  with  any  window  will
     clear  and  repaint  the screen from scratch.  If the window
     argument to wrefresh() is curscr, the screen in  immediately
     cleared  and  repainted from scratch.  This is how most pro-
     grams would implement  a  "repaint-screen"  function.   More
     information  on  using  curscr  is provided where its use is
     appropriate.

     The environment variables LINES and COLUMNS may  be  set  to
     override curses's idea of how large a screen is.

     If the environment variable TERMINFO is defined, any program
     using  curses  will  check  for  a local terminal definition
     before checking in the standard place.  For example, if  the
     environment  variable  TERM is set to sun, then the compiled
     terminal        definition        is        found         in



Sun Release 4.1   Last change: 21 January 1990                  4






CURSES(3V)             C LIBRARY FUNCTIONS             CURSES(3V)



     /usr/share/lib/terminfo/s/sun.   The  s  is  copied from the
     first letter of sun to avoid creation of huge  directories.)
     However,  if  TERMINFO  is set to $HOME/myterms, curses will
     first check $HOME/myterms/s/sun, and, if  that  fails,  will
     then  check  /usr/share/lib/terminfo/s/sun.   This is useful
     for developing experimental definitions or when  write  per-
     mission on /usr/share/lib/terminfo is not available.

     The  integer  variables  LINES  and  COLS  are  defined   in
     <curses.h>, and will be filled in by initscr() with the size
     of the screen.  For more  information,  see  the  subsection
     Terminfo-Level  Manipulations.  The constants TRUE and FALSE
     have the values 1 and 0, respectively.   The  constants  ERR
     and OK are returned by routines to indicate whether the rou-
     tine  successfully  completed.   These  constants  are  also
     defined in <curses.h>.

ROUTINES
     Many of the following routines have two  or  more  versions.
     The routines prefixed with w require a _w_i_n_d_o_w argument.  The
     routines prefixed with p  require  a  _p_a_d  argument.   Those
     without a prefix generally use stdscr.

     The routines prefixed with mv require _y and _x coordinates to
     move  to  before  performing the appropriate action.  The mv
     routines imply a call to move() before the call to the other
     routine.  The window argument is always specified before the
     coordinates. _y always refers to the row (of the window), and
     _x  always  refers  to  the column.  The upper left corner is
     always (0,0), not (1,1).  The  routines  prefixed  with  mvw
     take both a _w_i_n_d_o_w argument and _y and _x coordinates.

     In each case, _w_i_n is the window affected and _p_a_d is the  pad
     affected.   (win  and  _p_a_d  are  always  of  type WINDOW *.)
     Option-setting routines require a boolean flag _b_f  with  the
     value TRUE or FALSE.  (_b_f is always of type bool.) The types
     WINDOW, bool, and chtype  are  defined  in  <curses.h>  (see
     SYNOPSIS for a summary of what types all variables are).

     All routines return either the integer ERR  or  the  integer
     OK,  unless  otherwise noted.  Routines that return pointers
     always return NULL on error.

  Overall Screen Manipulation
     WINDOW *initscr()   The first routine called  should  almost
                         always be initscr().  The exceptions are
                         slk_init(), filter(), and  ripoffline().
                         This  will  determine  the terminal type
                         and initialize all  curses  data  struc-
                         tures.  initscr() also arranges that the
                         first call to refresh() will  clear  the
                         screen.  If errors occur, initscr() will



Sun Release 4.1   Last change: 21 January 1990                  5






CURSES(3V)             C LIBRARY FUNCTIONS             CURSES(3V)



                         write an appropriate  error  message  to
                         standard  error  and  exit; otherwise, a
                         pointer to stdscr is returned.   If  the
                         program  wants  an  indication  of error
                         conditions,  newterm()  should  be  used
                         instead  of initscr().  initscr() should
                         only be called once per application.

     endwin()            A program should  always  call  endwin()
                         before  exiting  or escaping from curses
                         mode temporarily, to do a  shell  escape
                         or  system(3)  call,  for example.  This
                         routine will  restore  termio(4)  modes,
                         move the cursor to the lower left corner
                         of the screen  and  reset  the  terminal
                         into  the  proper  non-visual  mode.  To
                         resume after a  temporary  escape,  call
                         wrefresh() or doupdate().

  Window and Pad Manipulation
     refresh()

     wrefresh (_w_i_n)      These    routines    (or     prefresh(),
                         pnoutrefresh(), wnoutrefresh(), or doup-
                         date()) must be called to  write  output
                         to  the terminal, as most other routines
                         merely   manipulate   data   structures.
                         wrefresh()  copies  the  named window to
                         the  physical  terminal  screen,  taking
                         into  account  what  is already there in
                         order to minimize the amount of informa-
                         tion that's sent to the terminal (called
                         optimization).  refresh() does the  same
                         thing,   except  it  uses  stdscr  as  a
                         default window.   Unless  leaveok()  has
                         been enabled, the physical cursor of the
                         terminal is left at the location of  the
                         window's  cursor.  The number of charac-
                         ters output to the terminal is returned.

                         Note: refresh() is a macro.

     WINDOW *newwin (_n_l_i_n_e_s, _n_c_o_l_s, _b_e_g_i_n__y, _b_e_g_i_n__x)
                         Create and return a  pointer  to  a  new
                         window  with  the  given number of lines
                         (or rows), _n_l_i_n_e_s, and  columns,  _n_c_o_l_s.
                         The  upper  left corner of the window is
                         at line  _b_e_g_i_n__y,  column  _b_e_g_i_n__x.   If
                         either  _n_l_i_n_e_s  or _n_c_o_l_s is 0, they will
                         be set to the value of lines-_b_e_g_i_n__y and
                         cols-_b_e_g_i_n__x.   A new full-screen window
                         is created by calling newwin(0,0,0,0).



Sun Release 4.1   Last change: 21 January 1990                  6






CURSES(3V)             C LIBRARY FUNCTIONS             CURSES(3V)



     mvwin (_w_i_n, _y, _x)   Move the window so that the  upper  left
                         corner  will  be at position (_y, _x).  If
                         the move would cause the  window  to  be
                         off  the  screen, it is an error and the
                         window is not moved.

     WINDOW *subwin (_o_r_i_g, _n_l_i_n_e_s, _n_c_o_l_s, _b_e_g_i_n__y, _b_e_g_i_n__x)
                         Create and return a  pointer  to  a  new
                         window  with  the  given number of lines
                         (or rows), _n_l_i_n_e_s, and  columns,  _n_c_o_l_s.
                         The  window  is  at  position ( _b_e_g_i_n__y,
                         _b_e_g_i_n__x) on the screen.   This  position
                         is  relative  to  the screen, and not to
                         the window _o_r_i_g.  The window is made  in
                         the  middle  of the window _o_r_i_g, so that
                         changes made to one window  will  affect
                         both  windows.  When using this routine,
                         often  it  will  be  necessary  to  call
                         touchwin() or touchline() on _o_r_i_g before
                         calling wrefresh.

      delwin (_w_i_n)       Delete the named window, freeing up  all
                         memory  associated with it.  In the case
                         of   overlapping   windows,   subwindows
                         should  be  deleted before the main win-
                         dow.

  Output
     These routines are used to "draw" text on windows.


     addch (_c_h)

     waddch (_w_i_n, _c_h)

     mvaddch (_y, _x, _c_h)

     mvwaddch (_w_i_n, _y, _x, _c_h)
                         The character _c_h is put into the  window
                         at  the  current  cursor position of the
                         window and the position  of  the  window
                         cursor  is  advanced.   Its  function is
                         similar  to  that  of   putchar()   (see
                         putc(3s)).   At  the  right  margin,  an
                         automatic newline is performed.  At  the
                         bottom   of  the  scrolling  region,  if
                         scrollok()  is  enabled,  the  scrolling
                         region will be scrolled up one line.

                         If _c_h is a TAB, NEWLINE,  or  backspace,
                         the  cursor  will be moved appropriately
                         within the window.  A NEWLINE also  does
                         a   clrtoeol()   before   moving.    TAB



Sun Release 4.1   Last change: 21 January 1990                  7





CURSES(3V)             C LIBRARY FUNCTIONS             CURSES(3V)



                         characters are considered to be at every
                         eighth column.  If _c_h is another control
                         character,  it  will  be  drawn  in  the
                         CTRL-X notation.  (Calling winch() after
                         adding  a  control  character  will  not
                         return   the   control   character,  but
                         instead will return  the  representation
                         of the control character.)

                         Video attributes can be combined with  a
                         character by or-ing them into the param-
                         eter.  This will result in these  attri-
                         butes  also  being set.  The intent here
                         is that text, including attributes,  can
                         be  copied  from  one  place  to another
                         using   inch()   and    addch().     See
                         standout(), below.

                         Note: _c_h is actually of type chtype, not
                         a character.

                         Note: addch(), mvaddch(), and mvwaddch()
                         are macros.



     addstr (_s_t_r)

     waddstr (_w_i_n, _s_t_r)

     mvwaddstr (_w_i_n, _y, _x, _s_t_r)

     mvaddstr (_y, _x, _s_t_r)
                         These routines write all the  characters
                         of  the null-terminated character string
                         _s_t_r  on  the  given  window.   This   is
                         equivalent  to calling waddch() once for
                         each character in the string.

                         Note:    addstr(),    mvaddstr(),    and
                         mvwaddstr() are macros.

     box (_w_i_n, _v_e_r_t_c_h, _h_o_r_c_h)
                         A box is drawn around the  edge  of  the
                         window,  _w_i_n.   _v_e_r_t_c_h and _h_o_r_c_h are the
                         characters the box is to be drawn  with.
                         If   _v_e_r_t_c_h   and   _h_o_r_c_h  are  0,  then
                         appropriate     default      characters,
                         ACS_VLINE and ACS_HLINE, will be used.

                         Note: _v_e_r_t_c_h and _h_o_r_c_h are  actually  of
                         type chtype, not characters.



Sun Release 4.1   Last change: 21 January 1990                  8






CURSES(3V)             C LIBRARY FUNCTIONS             CURSES(3V)



     erase()

     werase (_w_i_n)        These  routines  copy  blanks  to  every
                         position in the window.

                         Note: erase() is a macro.

     clear()

     wclear (_w_i_n)        These  routines  are  like  erase()  and
                         werase(),  but they also call clearok(),
                         arranging  that  the  screen   will   be
                         cleared  completely  on the next call to
                         wrefresh()   for   that   window,    and
                         repainted from scratch.

                         Note: clear() is a macro.

     clrtobot()

     wclrtobot (_w_i_n)     All lines below the cursor in this  win-
                         dow  are erased.  Also, the current line
                         to the right of the  cursor,  inclusive,
                         is erased.

                         Note: clrtobot() is a macro.

     clrtoeol()

     wclrtoeol (_w_i_n)     The current line to  the  right  of  the
                         cursor, inclusive, is erased.

                         Note: clrtoeol() is a macro.

     delch()

     wdelch (_w_i_n)

     mvdelch (_y, _x)

     mvwdelch (_w_i_n, _y, _x)
                         The character under the  cursor  in  the
                         window  is  deleted.   All characters to
                         the right on the same line are moved  to
                         the left one position and the last char-
                         acter on  the  line  is  filled  with  a
                         blank.   The  cursor  position  does not
                         change  (after  moving  to  (_y,  _x),  if
                         specified).   This does not imply use of
                         the hardware "delete-character" feature.

                         Note: delch(), mvdelch(), and mvwdelch()



Sun Release 4.1   Last change: 21 January 1990                  9






CURSES(3V)             C LIBRARY FUNCTIONS             CURSES(3V)



                         are macros.

     deleteln()

     wdeleteln (_w_i_n)     The line under the cursor in the  window
                         is deleted.  All lines below the current
                         line are moved up one line.  The  bottom
                         line of the window is cleared.  The cur-
                         sor position does not change.  This does
                         not  imply  use of the hardware "delete-
                         line" feature.

                         Note: deleteln() is a macro.

     getyx (_w_i_n, _y, _x)   The cursor position  of  the  window  is
                         placed  in  the  two integer variables _y
                         and _x.  This is implemented as a  macro,
                         so  no `&' is necessary before the vari-
                         ables.

     insch (_c_h)

     winsch (_w_i_n, _c_h)

     mvwinsch (_w_i_n, _y, _x, _c_h)

     mvinsch (_y, _x, _c_h)  The character _c_h is inserted before  the
                         character under the cursor.  All charac-
                         ters to the right are moved one SPACE to

usr/src/tin-1.18/curses.c
/*
 *  Project   : tin - a threaded Netnews reader
 *  Module    : curses.c
 *  Author    : D.Taylor & I.Lea
 *  Created   : 01-01-86
 *  Updated   : 11-10-92
 *  Notes     : This is a screen management library borrowed with permission
 *              from the Elm mail system (a great mailer--I highly recommend
 *              it!).This library was hacked to provide what tin needs.
 *  Copyright : Copyright (c) 1986-92 Dave Taylor & Iain Lea
 *              The Elm Mail System  -  $Revision: 2.1 $   $State: Exp $
 */

#ifndef ns32000
#	undef	sinix
#endif

#include <stdio.h>
#ifndef	AMIGA
#	include <curses.h>
#	include <sys/errno.h>
#else
#	include <errno.h>
#	define	TRUE 	1
#	define	FALSE 	0
#endif

#define DEFAULT_LINES_ON_TERMINAL	24
#define DEFAULT_COLUMNS_ON_TERMINAL	80

int LINES = 23;
int COLS  = 80;
int inverse_okay = TRUE;
static int _inraw = FALSE;	/* are we IN rawmode?    */
int _hp_glitch = FALSE;		/* standout not erased by overwriting on HP terms */

#ifndef INDEX_DAEMON

#define		BACKSPACE	'\b'
#define		VERY_LONG_STRING	2500

#if defined(AMIGA) || defined(BSD)
#	ifndef BSD4_1
#		include <sgtty.h>
#	else
#		include <termio.h>
#	endif
#else
#	ifndef SYSV
#		ifndef MINIX
#			ifdef sinix
#				include <termios.h>
#			else
#				include <termio.h>
#			endif
#		else
#			include <sgtty.h>
#		endif
#	else
#		ifdef __hpux
#			include <termio.h>
#		endif
#	endif
#endif

#define TTYIN	0

#ifdef SHORTNAMES
# define _clearinverse	_clrinv
# define _cleartoeoln	_clrtoeoln
# define _cleartoeos	_clr2eos
#endif

#if defined(AMIGA) || defined(BSD) || defined(MINIX)
#	ifdef TCGETA
#		undef TCGETA
#	endif
#	define TCGETA	TIOCGETP
#	ifdef TCSETAW
#		undef TCSETAW
#	endif
#	define TCSETAW	TIOCSETP
struct sgttyb _raw_tty,
	      _original_tty;
#else
#	ifdef sinix
#		ifndef TCGETA
#			define TCGETA	STCGETA
#		endif
#		ifndef TCSETA
#			define TCSETAW	STCSETAW
#		endif
struct termios _raw_tty, 
              _original_tty;
#	else
struct termio _raw_tty, 
              _original_tty;
#	endif
#endif

static char *_clearscreen, *_moveto, *_cleartoeoln, *_cleartoeos,
			*_setinverse, *_clearinverse, *_setunderline, *_clearunderline,
			*_terminalinit, *_terminalend, *_keypadlocal, *_keypadxmit;

static int _lines,_columns;

#ifndef AMIGA
static char _terminal[1024];              /* Storage for terminal entry */
static char _capabilities[1024];           /* String for cursor motion */

static char *ptr = _capabilities;	/* for buffering         */

extern char	*tgetstr ();		/* Get termcap capability (string) */
extern char	*tgoto ();		/* and the goto stuff    */
extern int	tgetflag ();		/* Get termcap capability (boolean) */
extern int	tgetnum ();		/* Get termcap capability (number) */
#endif	/* AMIGA */

static int in_inverse;			/* 1 when in inverse, 0 otherwise */
int	outchar ();			/* char output for tputs */

#endif /* INDEX_DAEMON */

#include "tin.h"


void setup_screen ()
{
	/*
	 * get screen size from termcap entry & setup sizes
	 */
	ScreenSize (&LINES, &COLS);
	cmd_line = FALSE;
	Raw (TRUE);
	set_win_size (&LINES, &COLS);
}


#ifndef AMIGA

int InitScreen ()
{
#ifndef INDEX_DAEMON

	extern int tgetent();      /* get termcap entry */
	char termname[40], *p;
	
	if ((p = (char *) getenv ("TERM")) == NULL) {
		fprintf (stderr, "%s: TERM variable must be set to use screen capabilities\n", progname);
		return (FALSE);
	}
	if (strcpy (termname, p) == NULL) {
		fprintf (stderr,"%s: Can't get TERM variable\n", progname);
		return (FALSE);
	}
	if (tgetent (_terminal, termname) != 1) {
		fprintf (stderr,"%s: Can't get entry for TERM\n", progname);
		return (FALSE);
	}

	/* load in all those pesky values */
	_clearscreen    = tgetstr ("cl", &ptr);
	_moveto         = tgetstr ("cm", &ptr);
	_cleartoeoln    = tgetstr ("ce", &ptr);
	_cleartoeos     = tgetstr ("cd", &ptr);
	_lines          = tgetnum ("li");
	_columns        = tgetnum ("co");
	_setinverse     = tgetstr ("so", &ptr);
	_clearinverse   = tgetstr ("se", &ptr);
	_setunderline   = tgetstr ("us", &ptr);
	_clearunderline = tgetstr ("ue", &ptr);
	_terminalinit   = tgetstr ("ti", &ptr);
	_terminalend    = tgetstr ("te", &ptr);
	_keypadlocal    = tgetstr ("ke", &ptr);
	_keypadxmit     = tgetstr ("ks", &ptr);

	_hp_glitch = tgetflag ("xs");

	InitWin ();

	if (!_clearscreen) {
		fprintf (stderr,
			"%s: Terminal must have clearscreen (cl) capability\n",progname);
		return (FALSE);
	}
	if (!_moveto) {
		fprintf (stderr,
			"%s: Terminal must have cursor motion (cm)\n", progname);
		return (FALSE);
	}
	if (!_cleartoeoln) {
		fprintf (stderr,
			"%s: Terminal must have clear to end-of-line (ce)\n", progname);
		return (FALSE);
	}
	if (!_cleartoeos) {
		fprintf (stderr,
			"%s: Terminal must have clear to end-of-screen (cd)\n", progname);
		return (FALSE);
	}
	if (_lines == -1)
		_lines = DEFAULT_LINES_ON_TERMINAL;
	if (_columns == -1)
		_columns = DEFAULT_COLUMNS_ON_TERMINAL;
	/* 
	 * kludge to workaround no inverse 
	 */
	if (_setinverse == 0) {
		_setinverse = _setunderline;
		_clearinverse = _clearunderline;
		if (_setinverse == 0)
			draw_arrow_mark = 1;
	}
	return (TRUE);

#else

	return (FALSE);

#endif /* INDEX_DAEMON */
}

#else	/* AMIGA InitScreen() */

int InitScreen ()
{
#ifndef			INDEX_DAEMON

	char c,*ptr,buf[32];

	/* 
	 * we're going to assume a terminal here... 
	 */

	_clearscreen	= "\033[1;1H\033[J";
	_moveto		= "\033[%d;%dH";	/* not a termcap string! */
	_cleartoeoln	= "\033[K";
	_cleartoeos	= "\033[J";

	_setinverse	= "\033[7m";
	_clearinverse	= "\033[0m";
	_setunderline	= "\033[4m";
	_clearunderline	= "\033[0m";
	_terminalinit	= "\033c";
	_terminalend	= "\033c";
	_keypadlocal	= "";
	_keypadxmit	= "";

	InitWin ();

	_lines = _columns = -1;

	/* 
	 * Get lines and columns from environment settings - useful when
         * you're using something other than an Amiga window 
         */
         
	if (ptr = getenv("LINES")) {
		_lines = atol(ptr);
	}
	if (ptr = getenv("COLUMNS")) {
		_columns = atol(ptr);
	}

	/* 
	 * If that failed, try get a response from the console itself 
	 */

	if (_lines == -1 || _columns == -1) {
		Raw (TRUE);

		tputs ("\2330 q",1,outchar);	/* identify yourself */
		fflush (stdout);

getsize:
		while (ReadCh () != 0x9b) {
			;	/* Look for escape */
		}
		/* get top */
		ptr = buf;
		do {	
			c = *ptr++ = ReadCh ();
		} while (isdigit(c));

		if (c != ';') { 
			goto getsize;
		}
		
		/* get right */
		ptr = buf;
		do {	
			c = *ptr++ = ReadCh ();
		} while (isdigit(c));

		if (c != ';') { 
			goto getsize;
		}
		
		/* get bottom */
		ptr = buf;
		do {	
			c = *ptr++ = ReadCh ();
		} while (isdigit(c));

		if (c != ';') {
			goto getsize;
		}

		*ptr = 0;
		_lines = atol (buf);

		/* get right */
		ptr = buf;
		do {	
			c = *ptr++ = ReadCh ();
		} while (isdigit(c));

		if (c != ' ') {
			goto getsize;
		}
		if (ReadCh () != 'r') { 
			goto getsize;
		}

		*ptr = 0;
		_columns = atol (buf);
	}

	Raw (FALSE);

	return (TRUE);
#else
	return (FALSE);

#endif /* INDEX_DAEMON */
}

#endif	/* AMIGA */

/*
 *  returns the number of lines and columns on the display.
 */
 
void ScreenSize (num_lines, num_columns)
	int *num_lines, *num_columns;
{
#ifndef INDEX_DAEMON

	if (_lines == 0) _lines = DEFAULT_LINES_ON_TERMINAL;
	if (_columns == 0) _columns = DEFAULT_COLUMNS_ON_TERMINAL;

	*num_lines = _lines - 1;		/* assume index from zero*/
	*num_columns = _columns;		/* assume index from one */

#endif /* INDEX_DAEMON */
}

void InitWin ()
{
#ifndef INDEX_DAEMON

	if (_terminalinit) {
		tputs (_terminalinit, 1, outchar);
		fflush (stdout);
	}	
	set_keypad_on ();

#endif /* INDEX_DAEMON */
}

void EndWin ()
{
#ifndef INDEX_DAEMON

	if (_terminalend) {
		tputs (_terminalend, 1, outchar);
		fflush (stdout);
	}
	set_keypad_off ();
	
	
#endif /* INDEX_DAEMON */
}

void set_keypad_on ()
{
#ifndef INDEX_DAEMON
#    ifdef HAVE_KEYPAD
 	if (use_keypad && _keypadxmit) {
		tputs (_keypadxmit, 1, outchar);
		fflush (stdout);
	}
#    endif
#endif /* INDEX_DAEMON */
}

void set_keypad_off ()
{
#ifndef INDEX_DAEMON
#    ifdef HAVE_KEYPAD
	if (use_keypad && _keypadlocal) {
		tputs (_keypadlocal, 1, outchar);
		fflush (stdout);
	}
#    endif
#endif /* INDEX_DAEMON */
}

/*
 *  clear the screen: returns -1 if not capable
 */

void ClearScreen ()
{
#ifndef INDEX_DAEMON

	tputs (_clearscreen, 1, outchar);
	fflush (stdout);      /* clear the output buffer */

#endif /* INDEX_DAEMON */
}

/*
 *  move cursor to the specified row column on the screen.
 *  0,0 is the top left!
 */

#ifndef AMIGA

void MoveCursor (row, col)
	int row, col;
{
#ifndef INDEX_DAEMON

	char *stuff, *tgoto();

	stuff = tgoto (_moveto, col, row);
	tputs (stuff, 1, outchar);
	fflush (stdout);

#endif /* INDEX_DAEMON */
}

#else	/* AMIGA MoveCursor() */

void MoveCursor (row, col)
	int row, col;
{
#ifndef INDEX_DAEMON

	char stuff[12], *tgoto();

	sprintf (stuff,_moveto, row+1, col+1);
	tputs (stuff, 1, outchar);
	fflush (stdout);

#endif /* INDEX_DAEMON */
}

#endif	/* AMIGA */

/*
 *  clear to end of line
 */

void CleartoEOLN ()
{
#ifndef INDEX_DAEMON

	tputs (_cleartoeoln, 1, outchar);
	fflush (stdout);  /* clear the output buffer */

#endif /* INDEX_DAEMON */
}

/*
 *  clear to end of screen
 */

void CleartoEOS ()
{
#ifndef INDEX_DAEMON

	int i;
	
	if (_cleartoeos) {
		tputs (_cleartoeos, 1, outchar);
	} else {
		for (i=_lines ; i < _lines ; i++) {
			MoveCursor (i, 0);
			CleartoEOLN ();
		}
	}
	fflush (stdout);  /* clear the output buffer */

#endif /* INDEX_DAEMON */
}

/*
 *  set inverse video mode
 */

void StartInverse ()
{
#ifndef INDEX_DAEMON

	in_inverse = 1;
	if (_setinverse && inverse_okay)
		tputs (_setinverse, 1, outchar);
	fflush (stdout);

#endif /* INDEX_DAEMON */
}

/*
 *  compliment of startinverse
 */

void EndInverse ()
{
#ifndef INDEX_DAEMON

	in_inverse = 0;
	if (_clearinverse && inverse_okay)
		tputs (_clearinverse, 1, outchar);
	fflush (stdout);

#endif /* INDEX_DAEMON */
}

/*
 *  toggle inverse video mode
 */

void ToggleInverse ()
{
#ifndef INDEX_DAEMON

	if (in_inverse == 0)
		StartInverse();
	else
		EndInverse();

#endif /* INDEX_DAEMON */
}

/*
 *  returns either 1 or 0, for ON or OFF
 */

int RawState()
{
	return (_inraw);
}

/*
 *  state is either TRUE or FALSE, as indicated by call
 */

void Raw(state)
	int state;
{
#ifndef INDEX_DAEMON

	if (state == FALSE && _inraw) {
	  (void) ioctl(TTYIN, TCSETAW, &_original_tty);
	  _inraw = 0;
	} else if (state == TRUE && ! _inraw) {
	  (void) ioctl(TTYIN, TCGETA, &_original_tty);	/** current setting **/
	  (void) ioctl(TTYIN, TCGETA, &_raw_tty);    /** again! **/

#if defined(AMIGA) || defined(BSD) || defined(MINIX)
	  _raw_tty.sg_flags &= ~(ECHO | CRMOD);	/* echo off */
	  _raw_tty.sg_flags |= CBREAK;		/* raw on */
#ifdef AMIGA	
	  _raw_tty.sg_flags |= RAW; /* Manx-C 5.2 does not support CBREAK */
#endif
#else
	  _raw_tty.c_lflag &= ~(ICANON | ECHO);	/* noecho raw mode        */

	  _raw_tty.c_cc[VMIN] = '\01';	/* minimum # of chars to queue    */
	  _raw_tty.c_cc[VTIME] = '\0';	/* minimum time to wait for input */
#endif

	  (void) ioctl(TTYIN, TCSETAW, &_raw_tty);
	  _inraw = 1;
	}
	
#endif /* INDEX_DAEMON */
}

/*
 *  read a character with Raw mode set!
 */

int ReadCh()
{
#ifndef INDEX_DAEMON
	extern int errno;
	char ch;
	register int result = 0;
	
#ifdef READ_CHAR_HACK
#undef getc
	while ((result = getc(stdin)) == EOF) {
		if (feof(stdin))
			break;

#ifdef EINTR
		if (ferror(stdin) && errno != EINTR)
#else
		if (ferror(stdin))
#endif
			break;

		clearerr(stdin);
	}

	return ((result == EOF) ? EOF : result & 0xFF);
#else
#  ifdef EINTR
	while ((result = read (0, &ch, 1)) < 0 && errno == EINTR)
		;	/* spin on signal interrupts */
#  else
	result = read (0, &ch, 1);
#  endif
        return((result <= 0 ) ? EOF : ch & 0xFF);
#endif		

#endif /* INDEX_DAEMON */
}

/*
 *  output a character. From tputs... (Note: this CANNOT be a macro!)
 */

int outchar (c)
	int c;
{
	fputc (c, stdout);
}
